std/experimental/allocator/package.d

```diff
@@ -895,7 +895,17 @@ nothrow @safe @nogc unittest
     assertThrown(make!InvalidImpureStruct(Mallocator.instance, 42));
 }
 
-private void fillWithMemcpy(T)(void[] array, auto ref T filler) nothrow
+private void fillWithMemcpy(T)(scope void[] array, auto ref T filler) nothrow
+if (T.sizeof == 1)
+{
+    import core.stdc.string : memset;
+    import std.traits : CopyConstness;
+    if (!array.length) return;
+    memset(array.ptr, *cast(CopyConstness!(T*, ubyte*)) &filler, array.length);
```

Comment:
Why not `cast(int) filler` as the API expects?

```
pure nothrow @nogc @system void* memset(return void* s, int c, size_t n); 
```

https://dlang.org/phobos/core_stdc_string.html#.memset

For other reviewers are wondering too why `int` is required by the C header:

> The value is passed as an int, but the function fills the block of memory using the unsigned char conversion of this value.

http://www.cplusplus.com/reference/cstring/memset/

Reply:
> Why not `cast(int) filler`

My rationale is that it's possible for to T override opCast so `cast(int) filler` could return something whose low byte is not bitwise identical to T.init.

>as the API expects

ubyte will be promoted to int

Reply:
What about using a union to make the intent clearer, instead of that line noise complicated cast?
````
union U { T t; ubyte asByte; }
U u;
u.t = filler;
memset(array.ptr, u.filler, array.length);
````

Reply:
> `u.t = filler;`

I need to avoid `opAssign` in order for this function to always have the same behavior as the existing `fillWithMemcpy` which never calls `opAssign`. I could do:

```d
union U { Unqual!T t; ubyte asByte; }
U u = void;
memcpy(&u.t, &filler, 1);
memset(array.ptr, u.asByte, array.length);
```
if you think that would be clearer.

Reply:
Hmm. If you wish to avoid opAssign, I suppose there's really no other way than to dereference a pointer cast to `ubyte`. Fair enough.